### DISCUSSION TITLE
Nightscout - Google Cloud  - Free DNS issue

### DISCUSSION THREAD

**Discussion by koscica** [NightscoutFoundation/xDrip#2788] [General]
Hi hello I have set up nightscout on Google Cloud and it has been woeking fine for a couple of months. However today I keep getting 502 Bad Gateway error.  I noticed that freedns.afraid.org site is offline. Any suggestions?

**Comment by koscica** [comment]
restarted nightscout on google cloud and now it is working

**Comment by kresimirjedud** [comment]
I have the same Issue, restartted GC. It worked for a while and then again
BAD GATEWAY ... I think we have some global issue ... Not sure what it is
...

*Regards,*

*KJ*


sri, 19. tra 2023. u 11:15 koscica ***@***.***> napisao je:

**Comment by koscica** [comment]
yep same thing again

**Comment by BapLL22** [comment]
Same thing here :(

**Comment by tzachi-dar** [comment]
It seems that this is happening to a lot of machines. Here is what seems to solve the problem (only steps 1-4 are actually needed to solve the issue, the other ones are if you are still having problems.):

1) connect using ssh to the machine.
2) From the menu select the last option (exit to shell).
3) please run uptime. (This step is not a must, but from what I saw a lot of machines have been rebooted, so it is interesting if your machine has also rebooted). 
4) From the shell, please run:
```
 sudo screen -L -Logfile /tmp/screen.log  -dmS nightscout sudo -u nobody bash /etc/nightscout-start.sh
```
After 30 seconds the issue should be solved.
5) (not a must, for debug only) please run:
```
sudo ps -ef | grep -i node
```
and make a record of the output. Should be something like:
```
nobody      1210    1201  1 08:49 pts/1    00:01:16 node server.js
g6610mu+    1693    1166  0 10:54 pts/0    00:00:00 grep --color=auto -i node
```
The interesting number to remember is the 1210.
6) (only if there is a problem) Please let us know what the problem is, and what is the result of 
```
tail -n 100 /tmp/screen.log
```

**Reply by koscica** [reply]
thx mate.
 worked like a charm on nightscout  installations

**Reply by BapLL22** [reply]
Thx ! It works well now

**Comment by mironmat** [comment]
HI and thx @tzachi-dar . 
Any clue why the scrip crashes after reboot? it is part of rc.local, looks like the same one without "logging" part.

**Reply by Navid200** [reply]
Do you mean you are still having problems?

**Reply by mironmat** [reply]
before this solution (shared on FB by Navid Fo, i guess by you :) ), i was doing my own troubleshooting and i tried to fire command: "sudo bash /etc/nightscout-start.sh". But NS worked for 5mins and the same issue. Now just used command from here and still checking, but so far so good. But just wondering if I would need to lauch this again after server reboot. I guess yes, if rc.local part is not working. But we will see. For now, it's working ok, so again big thx.

**Reply by Navid200** [reply]
I woke up and saw all these messages.
I have had no issue with my setup.  It is possible that those experiecing this have something in common.  It will help to find out what it is if that's the case.

Do you mind posting a screenshot of your status page?  That page has nothing private shown.  But, it will show me exactly what you are running.
https://navid200.github.io/xDrip/docs/Nightscout/Status.html

**Reply by mironmat** [reply]
np, happy to do some testing, share logs etc. I did update update today (apt-get update/upgrade), hoping it will resolve the issue, but not. Here is screen: 
![image](https://user-images.githubusercontent.com/26772702/233092797-f289d60c-6141-47d3-be23-0cb2529f2f48.png)

**Reply by mironmat** [reply]
so the issue was on ubuntu: 20.04.05 as well as on 20.04.06. I have different port, but it shouldn't be a prb. and using my prv. DNS name. 
![image](https://user-images.githubusercontent.com/26772702/233093624-e70ee679-7305-4f6c-a295-87140145eb53.png)

**Reply by Navid200** [reply]
Has this been red all along since you created the machine?
![233092797-f289d60c-6141-47d3-be23-0cb2529f2f48](https://user-images.githubusercontent.com/51497406/233093802-cc58ca41-4b1a-4e99-a3d7-42feaa060ae9.png)

**Reply by mironmat** [reply]
no, today morning it was black/normal. As mentioned above, did system update after issue, so i guess it's red as ubuntu version is not 20.04.5

**Reply by koscica** [reply]
Here is my status

<img width="403" alt="image" src="https://user-images.githubusercontent.com/45534014/233094579-cc3f68b8-fe0b-41d3-a418-e958a25bccae.png">

**Reply by koscica** [reply]
I seem to have missing node modules. But it seems to be working fine

**Reply by Navid200** [reply]
There is no way in hell we can support all sort of systems people have decided to create deviating from our instructions.

If you have ignored our guide and have done something else, and you choose to post in a public forum stating that something is wrong with it, the least you can do is to state you you have not followed our guide.

I am curious how you do your job.  How would you deal with someone who complains about your job when they in fact are the cause of the problem.

I need a break.

**Reply by nlusternik** [reply]
I installed only yesterday for the first time and had 502 bad gateway today.
Ubuntu: 20.04.5 LTS.
Nothing red in the status.
The instructions above solved the problem.
![status](https://user-images.githubusercontent.com/130162890/233095838-454f801d-3746-4175-97e2-10e864f35e33.png)

**Reply by koscica** [reply]
> There is no way in hell we can support all sort of systems people have decided to create deviating from our instructions.
> 
> If you have ignored our guide and have done something else, and you choose to post in a public forum stating that something is wrong with it, the least you can do is to state you you have not followed our guide.
> 
> I am curious how you do your job. How would you deal with someone who complains about your job when they in fact are the cause of the problem.
> 
> I need a break.

I have followed the guide. I have no idea why got the message

**Reply by Navid200** [reply]
If you look at the page about how to update your platform (https://navid200.github.io/xDrip/docs/Nightscout/NS_SyncExecutables.html), you will see an image that looks exactly like your status page.  And it tells you how to fix it.

**Reply by mironmat** [reply]
> There is no way in hell we can support all sort of systems people have decided to create deviating from our instructions.
> 
> If you have ignored our guide and have done something else, and you choose to post in a public forum stating that something is wrong with it, the least you can do is to state you you have not followed our guide.
> 
> I am curious how you do your job. How would you deal with someone who complains about your job when they in fact are the cause of the problem.
> 
> I need a break.

Take a break man. I fully understand you. But when it comes to my OS, as I mentioned, today morning all was black - in line with your guide. I guess that sth. was changed/updated automatically and that's why so many people start receiving issue with NS script. We all are very grateful for the work you are doing.

**Reply by Navid200** [reply]
Are you using your private domain instead of using FreeDNS?

**Reply by mironmat** [reply]
I used FreeDNS but decided to change it to prv., so the answer is yes, using prv. domain. So I understand that i have now custom config: prv. domain and updated OS, which means that it is not in line with your guide line, so not supported by you. So no worries man, I do not have any expectations. As mentioned above, if you are happy to test/check sth. happy to help. I guess lots of other users are not familiar with tech aspects. I have "basic knowledge" so could be helpful. But from the other side, such people like me make you headache :) fully understand.

**Reply by Navid200** [reply]
We have talked about supporting private domains.  The opinion I got, from a knowledgeable member of our team, was that was not practical.

The way I understand it, we have mechanisms in place that ensure the hostname is linked to the ip address.  This is done because we know how to access FreeDNS directly.  So, when you run install phase 2, a lot happens to ensure that everything is set up correctly.  So, after a reboot, everything comes up as it should.

There is no way we can provide the same for other domains because we don't know how to access those providers.
We can include a few.  For example, we can include a selection form where you can select if you use FreeDNS or if you use this or if you use that.  But, that means every time we are reviewing a new submission, we will have to test it for all of them.  That will make the job of maintaining this project harder than it needs to be.  I don't have a private domain.   So, how can I test it?

The only reason we have been given by people who have chosen to use a private domain is that they are worried their FreeDNS domain may be blocked.
I have gone over what has worked for many here: https://navid200.github.io/xDrip/docs/Nightscout/FD_Domains.html

I admit that this is not a perfect solution.  But, it has worked for many.
We are open to suggestions.

Anyone is free to use a private domain.  I have never asked people not to.  But, it is not something we can troubleshoot.  If you use a private domain, you may have to troubleshoot it yourself.

**Reply by BapLL22** [reply]
Here's my status
<img width="393" alt="Capture d’écran 2023-04-19 à 16 44 19" src="https://user-images.githubusercontent.com/127324594/233114025-fa635180-2d10-4a79-8e5b-bbf1bfd37311.png">

**Reply by BapLL22** [reply]
The instructions above also solved the problem.

**Comment by 1010bits** [comment]
The instructions by tzachi-dar https://github.com/NightscoutFoundation/xDrip/discussions/2788#discussioncomment-5659798 worked so far. Thanks! Note that the fix didn't survive a reboot.

Prior to the tzachi-dar fix, I updated the platform and ran the phase 1 install again. Before updating the platform, I think the status showed 2022.12.06 (I realize that date does not show in the list of [releases](https://navid200.github.io/xDrip/docs/Nightscout/GC_ReleaseNotes.html)).

P.S. Thanks so much for xDrip and for this Nightscout system! Our family depends on both systems every day! 
![2023-04-19 xDrip Nightscout status](https://user-images.githubusercontent.com/53198611/233127788-ac1ee9c3-2502-431c-ba02-f8ad08912bb0.jpg)

**Reply by Navid200** [reply]
> Note that the fix doesn't survive a reboot.

@1010bits 


Would you please reboot now?
I suspect now that FreeDNS is back up, it may work.

**Reply by Navid200** [reply]
If a reboot still fails, would you please show the content of the following file?

/etc/rc.local

**Reply by 1010bits** [reply]
Yes, I rebooted, and things are working fine. Thanks! I didn't realize that FreeDNS was having problems. 

Just in case it's useful, here is /etc/rc.local:
  GNU nano 4.8                                     /etc/rc.local                                               
#!/bin/bash

PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin"

cd /tmp

swapon /var/SWAP

service snapd stop

/usr/local/bin/noip2 &
service mongodb start

screen -dmS nightscout sudo -u nobody bash /etc/nightscout-start.sh

service nginx start

. /etc/free-dns.sh
wget -O /tmp/freedns.txt --retry-connrefused --waitretry=1 --read-timeout=20 --timeout=15 $DIRECTURL

**Reply by Navid200** [reply]
@tzachi-dar 
@1010bits 

That's very helpful.


The rc.local file is executed after a reboot.
You can see in it the swap file being set up.
In your case, you can see that it is setting up noip.  I don't see it in mine.  If you run phase 2, this will be removed.  But, I don't think there is a problem leaving it as is.

It starts mongo.

But, do you see the last line?
Do you know what $DIRECTURL is?

It is a direct link to your hostname using an address from FreeDNS.  But, when FreeDNS is down, this will fail.

So, I can now see how everything went wrong for so many people.

Tzachi and I will talk about this and may be change it so that this will not happen again the next time FreeDNS goes down.

I feel much better now.  I think I understand how everything got messed up for so many people.

**Reply by Navid200** [reply]
@1010bits Would you please do me another favor?
Would you please see if /user/local/bin/noip2 exists?

**Reply by 1010bits** [reply]
Good job finding that! I like your idea of modifying things to have a solution in place for the next connectivity problem!

Yes, I'm happy to help, and yes, the file exists:
$ stat /usr/local/bin/noip2
  File: /usr/local/bin/noip2
  Size: 91392          Blocks: 184        IO Block: 4096   regular file
Device: 801h/2049d     Inode: 4728        Links: 1
Access: (0755/-rwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2023-04-20 17:12:36.654132960 +0000
Modify: 2022-11-12 03:01:40.257916301 +0000
Change: 2022-11-12 03:01:40.257916301 +0000
 Birth: -

**Comment by tuim** [comment]
I have Ubuntu unattended update setup and the NightScout broke (502 error) after an OS update. The instructions above worked great. I can confirm that after the reboot the server works just for a little while and stops again. 
Thank you very much for the teams supporting this. I really appreciate it.

**Reply by Navid200** [reply]
We have not tested our platform after any updates to Ubuntu.  
How/why did you set up unattended updates?

**Reply by tuim** [reply]
Unattended Upgrades ensure that important security patches for installed packages are automatically downloaded and installed without needing any manual intervention from an end-user (hence the word unattended). You can see more details here:
https://www.kolide.com/features/checks/ubuntu-unattended-upgrades

**Reply by Navid200** [reply]
What security?  Why would that be a concern?
You don't use this setup for searching the web, do you?  What are your security concerns?  How is anyone going to cause a security issue for your setup if they don't have your API_SECRET password and your Google account password?

All of that is irrelevant.  If you are not running Ubuntu 20.04.5 and are running something else, we have not tested it.

We can talk about updating Ubuntu on our platform.  If there is a need, we will.
But, we are not going to update every package as soon as a new one comes out.

**Comment by maosuna** [comment]


**Reply by pirupic2** [reply]
Hello, It does not work for me either

ensuring index for: sgv
ensuring index for: mbg
ensuring index for: sysTime
ensuring index for: dateString
ensuring index for: [object Object]
ensuring index for: created_at
ensuring index for: eventType
ensuring index for: insulin
ensuring index for: carbs
ensuring index for: glucose
ensuring index for: enteredBy
ensuring index for: boluscalc.foods._id
ensuring index for: notes
ensuring index for: NSCLIENT_ID
ensuring index for: percent
ensuring index for: absolute
ensuring index for: duration
ensuring index for: [object Object]
ensuring index for: created_at
ensuring index for: NSCLIENT_ID
ensuring index for: startDate
ensuring index for: type
ensuring index for: position
ensuring index for: hidden
ensuring index for: created_at
Executing setupListeners
Executing setupBridge
Dexcom bridge not enabled
Executing setupMMConnect
MiniMed Connect not enabled
Executing finishBoot
tick 2023-04-20T07:28:40.417Z
Boot event processing completed
Security settings: INSECURE_USE_HTTP= true , SECURE_HSTS_HEADER= true
API KEY present, enabling API
enabling feature: careportal
enabling feature: food
enabling feature: boluscalc
enabling feature: bwp
enabling feature: cob
enabling feature: bgi
enabling feature: pump
enabling feature: openaps
enabling feature: rawbg
enabling feature: iob
enabling feature: upbat
enabling feature: cage
enabling feature: sage
enabling feature: basal
enabling feature: treatmentnotify
enabling feature: bgnow
enabling feature: delta
enabling feature: direction
enabling feature: timeago
enabling feature: devicestatus
enabling feature: errorcodes
enabling feature: profile
enabling feature: bolus
enabling feature: dbsize
enabling feature: runtimestate
enabling feature: ar2
ensuring index for: identifier
ensuring index for: srvModified
ensuring index for: isValid
ensuring index for: identifier
ensuring index for: srvModified
ensuring index for: isValid
ensuring index for: identifier
ensuring index for: srvModified
ensuring index for: isValid
ensuring index for: identifier
ensuring index for: srvModified
ensuring index for: isValid
ensuring index for: identifier
ensuring index for: srvModified
ensuring index for: isValid
ensuring index for: identifier
ensuring index for: srvModified
ensuring index for: isValid
Production environment detected, enabling Minify
WARNING: express-minify cache directory is not writeable, fallback to memory cache.
Listening on port 1337 127.0.0.1
events.js:377
      throw er; // Unhandled 'error' event
      ^

Error: listen EADDRINUSE: address already in use 127.0.0.1:1337
    at Server.setupListenHandle [as _listen2] (net.js:1331:16)
    at listenInCluster (net.js:1379:12)
    at doListen (net.js:1516:7)
    at processTicksAndRejections (internal/process/task_queues.js:83:21)
Emitted 'error' event on Server instance at:
    at emitErrorNT (net.js:1358:8)
    at processTicksAndRejections (internal/process/task_queues.js:82:21) {
  code: 'EADDRINUSE',
  errno: -98,
  syscall: 'listen',
  address: '127.0.0.1',
  port: 1337
}

**Reply by pirupic2** [reply]
after 5 minutes and a second reboot it started to work for me

**Comment by Navid200** [comment]
For anyone whose Nightscout is not working, would you please show a screenshot of your status page?
This is what I mean by the status page:  https://navid200.github.io/xDrip/docs/Nightscout/Status.html

**Comment by Navid200** [comment]
There is a line in our code that starts Nightscsout after a reboot.  If that line is missing, I need to know.  If a reboot stops Nightscout for you, please tell me.  But, please show  the status page.

If for some reason, your ip address had changed, I can see that FreeDNS having been down could have been the cause of the problem.   But, it is back up.
And, no, please don't do a fresh install.  Please help me see what is wrong so that we can fix it.

**Comment by Navid200** [comment]
If you ran "Install Nightscout phase 2" when FreeDNS was down, it is not surprising that things broke for you.

Phase 2 uses FreeDNS.  I suspect that if you ran it, the line that starts Nightscout is not correct in your case.  So, every time you reboot, Nightscsout does not start for you.

I will open a PR to check that FreeDNS is up when someone attempts to run phase 2.  If it is down, it should provide a warning and not execute.

Please help me identify the problem so that it will not happen again.

So now that FreeDNS is up and running, if you run phase 2, it will work fine and it will correct the Nightcout start line.

**Comment by tuim** [comment]
I have rebooted my NightScout Server today and everything is working just fine! Thanks.

**Reply by Navid200** [reply]
Would you please stay out of this thread if you are not having a problem and are not interested in our approach?  
We may have a disagreement.  You are welcome to open another thread and I am more than happy to have an open and constructive conversation.

**Reply by tuim** [reply]
Hey Navid200, I was just trying to help. I just reported back since my server was not working yesterday, most likely because FreeDNS was having issues. Thank you for your support and have a great day.

**Comment by Navid200** [comment]
I have been able to recreate the problem.

Simply a restart of the server when FreeDNS was down resulted in an error that was not handled gracefully resulting in Nightscout not starting.

I am putting together a PR to fix this so that it would not happen again in case FreeDNS goes down again.
I should be able to submit the PR in a day or two.

**Comment by Loax28** [comment]
Hey, 

don´t know if this is closed or i can post my problem here:

I tried to update to NS15. So I tried to update the platform, then I got the red comment in status: start up dependence on freeDNS + FreeDNS ID and pass.
Therefore I did in the terminal phase 2 in Google cloud setup.
But now I cannot access to my nightscout (Bad gateway).
In installation I at first used the first way were I keep my certificate and then the second where I redo it. But I didn't to the steps after if that is needed. 
I didn't do the first phase.
Thank you for any help.
image
This is written above, what I also tried:
image

**Comment by Navid200** [comment]
@Loax28 

Please open a new thread, for clarity for anyone who may find the thread through search, and I will do everything I can to help.

**Reply by Loax28** [reply]
Thank you so much. I posted it here: https://github.com/NightscoutFoundation/xDrip/discussions/3377
